pkg/container-collection/container-collection.go

```diff
@@ -24,28 +24,78 @@ package containercollection
 import (
 	"sync"
 
+	dockerclient "github.com/docker/docker/client"
+	"k8s.io/client-go/kubernetes"
+
 	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
 	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+	"github.com/kinvolk/inspektor-gadget/pkg/runcfanotify"
 )
 
 // ContainerCollection holds a set of containers. It can be embedded as an
 // anonymous struct to help other structs implement the ContainerResolver
 // interface. For this reason, some methods are namespaced with 'Container' to
 // make this clear.
 type ContainerCollection struct {
-	// Keys:   Container ID
-	// Values: pb.ContainerDefinition
+	// Keys:   containerID string
+	// Values: container   *pb.ContainerDefinition
 	containers sync.Map
 
 	// subs contains a list of subscribers of container events
 	pubsub *pubsub.GadgetPubSub
+
+	// containerEnrichers are functions that automatically add metadata
+	// upon AddContainer
+	containerEnrichers []func(container *pb.ContainerDefinition, initial bool)
+
+	// dockercli is the Docker API client to be used by
+	// WithDockerEnrichment
+	dockercli *dockerclient.Client
+
+	// client to talk to the k8s API server to get information about pods
+	k8sClient *kubernetes.Clientset
+
+	// runcNotifier to be used by WithRuncFanotify
+	runcNotifier *runcfanotify.RuncNotifier
+
+	// initialContainers is used during the initialization process to
+	// gather initial containers and then call the enrichers
+	initialContainers []*pb.ContainerDefinition
+
+	// initialized tells if ContainerCollectionInitialize has been called.
+	initialized bool
 }
 
+// ContainerCollectionOption are options to pass to
+// ContainerCollectionInitialize using the functional option code pattern.
+type ContainerCollectionOption func(*ContainerCollection)
+
 // ContainerCollectionInitialize initializes a ContainerCollection. It is
 // useful when ContainerCollection is embedded as an anonymous struct because
 // we don't use a contructor in that case.
-func (cc *ContainerCollection) ContainerCollectionInitialize() {
+func (cc *ContainerCollection) ContainerCollectionInitialize(options ...ContainerCollectionOption) {
+	if cc.initialized {
+		panic("ContainerCollectionInitialize already called")
+	}
+
+	// Call functional options. This might fetch initial containers.
 	cc.pubsub = pubsub.NewGadgetPubSub()
+	for _, o := range options {
+		o(cc)
+	}
+
+	// Consume initial containers that might have been fetched by
+	// functional options. This is done after all functional options have
+	// been called, so that cc.containerEnrichers is fully set up.
+	for _, container := range cc.initialContainers {
+		for _, enricher := range cc.containerEnrichers {
+			enricher(container, true)
+		}
+		cc.containers.Store(container.Id, container)
+	}
+	cc.initialContainers = nil
+
+	cc.initialized = true
```

Comment:
I don't think it is worth adding complexity for a case which is clearly a bug. In Gadget Tracer Manager, `ContainerCollectionInitialize` is called from `newServer()` and no other routines can call `ContainerCollectionInitialize()` at the same time because nobody has a reference to the `GadgetTracerManager` yet. Same thing in `LocalGadgetManager`.

Reply:
> because nobody has a reference to the GadgetTracerManager yet. Same thing in LocalGadgetManager

This is true but you're supposing that the only users of `ContainerCollection` will always be `GadgetTracerManager` & `LocalGadgetManager.`We could also argue that calling multiple times  `ContainerCollectionInitialize` is a bug and we don't have to implement any measures against that (i.e. just drop the `initialized` member). 

I don't remember why we decided to embed `ContainerCollection` in `GadgetTracerManager`. If this would be a normal member then we could have a `containercollection.New(options ...ContainerCollectionOption) *ContainerCollection` and we won't have to worry about `ContainerCollectionInitialize` being called multiple times. Perhaps that's a cleaner approach?

Reply:
> I don't remember why we decided to embed ContainerCollection in GadgetTracerManager.

I did it this way so that `GadgetTracerManager` implements the `Resolver` interface defined in `pkg/gadgets/interface.go`. This is so gadgets can call the different lookup methods such as `t.resolver.LookupMntnsByPod`.

But this could be changed to make it a normal `containercollection.New()` method as you suggest. We would just have to update the controller to something like that:

```diff
-factory.Initialize(tracerManager, mgr.GetClient())
+factory.Initialize(tracerManager.ContainerCollection, mgr.GetClient())
```

+ some other cleanups because ContainerCollection does not implement  `PublishEvent()` because it's not related to containers.

Should I do that in the same PR?

Reply:
> Should I do that in the same PR?

I think it's perfectly fine to move to another PR. This has already enough changes :D